gareth-cross/rviz_satellite
# Fix frame jitter by splitting map and fixed-frame transforms

Fixes #56:
The jitter exists if the frequency of the NavSatFix callback is lower than RViz' framerate and the transform update rate. This is usually the case, GPS often comes with only 1Hz, while RViz" fps are usually 30-60Hz, localization transforms have usually a high frequency (>50Hz).
It origins from unsynchronized transforms: The scene node's pose is only updated on callback, but in-between callbacks, the robot might move. I.e.: the map-repaint lags behind.
In order to synchronize this, we need to transform the scene-node on every render-frame with the latest information, as usual with RViz Display plugins rendering data that is attached to a tf-frame.

So we split the transform from the tile map to the fixed-frame into two transforms, created at two locations:
- transform the tile-map from the NavSatFix tf frame into the map frame (rationale: the tiles are rigidly attached to the world by the Mercator projection, this relationship is not subject to change, independent on the robots pose)
- transform the tile-map from the map-frame into the fixed frame with every render-callback

Along with this change, we:
- changed the sequence of events for updating texture and geometry
- improved validation and actions taken when updating properties
- did a minor general cleanup
- updated the demo.launch to reflect a more comlex scenario (base_link and map are not identical any more; this enables better debugging of the transforms)


**TODO**:
- squash commits

**Test Plan**
- confirm that all the properties and their error handlers are working as before
- confirm that there is no jitter left independent of the fixed frame set: 
  - set RViz' fixed frame to a robot-fixed frame, camera target frame to _map_
  - move the robot fast
  - set a low update rate of the gps_fix topic (1Hz)
  - set a high update rate of the robot's tf (100Hz)
- check that status is updated correctly
- check error reporting (missing tfs, missing gps_fix)
- check that we don't have new issues
- check that the map is positioned exactly as before (no offset/rotation)
- confirm that the transforms are working as expected
